In [ ]:
import pandas as pd
import numpy as np

In [ ]:
df = pd.read_csv("Diamonds_train.csv")

In [ ]:
df.shape

(43152, 11)

In [ ]:
df.head(50)

Unnamed: 0  carat        cut color clarity  depth  table  price     x  \
0            1   0.23      Ideal     E     SI2   61.5   55.0    326  3.95   
1            2   0.21    Premium     E     SI1   59.8   61.0    326  3.89   
2            3   0.23       Good     E     VS1    NaN   65.0    327  4.05   
3            4   0.29    Premium     I     VS2   62.4   58.0    334  4.20   
4            5   0.31       Good     J     SI2   63.3   58.0    335  4.34   
5            6   0.24  Very Good     J    VVS2   62.8   57.0    336  3.94   
6            7   0.24  Very Good     I    VVS1   62.3   57.0    336  3.95   
7            8   0.26  Very Good     H     SI1   61.9   55.0    337  4.07   
8            9   0.22       Fair     E     VS2   65.1   61.0    337  3.87   
9           10   0.23  Very Good     H     VS1   59.4   61.0    338  4.00   
10          11   0.30       Good     J     SI1   64.0   55.0    339  4.25   
11          12   0.23      Ideal     J     VS1   62.8   56.0    340  3.93   
12          13   0.22    Premium     F     SI1   60.4   61.0    342  3.88   
13          14   0.31      Ideal     J     SI2   62.2   54.0    344  4.35   
14          15   0.20    Premium     E     SI2   60.2   62.0    345  3.79   
15          16   0.32    Premium     E      I1   60.9   58.0    345  4.38   
16          17   0.30      Ideal     I     SI2   62.0   54.0    348  4.31   
17          18   0.30       Good     J     SI1   63.4   54.0    351  4.23   
18          19   0.30       Good     J     SI1   63.8   56.0    351  4.23   
19          20   0.30  Very Good     J     SI1   62.7   59.0    351  4.21   
20          21   0.30       Good     I     SI2   63.3   56.0    351  4.26   
21          22   0.23  Very Good     E     VS2   63.8   55.0    352  3.85   
22          23   0.23  Very Good     H     VS1   61.0   57.0    353  3.94   
23          24   0.31  Very Good     J     SI1   59.4   62.0    353  4.39   
24          25   0.31  Very Good     J     SI1   58.1   62.0    353  4.44   
25          26   0.23  Very Good     G    VVS2   60.4   58.0    354  3.97   
26          27   0.24    Premium     I     VS1   62.5   57.0    355  3.97   
27          28   0.30  Very Good     J     VS2   62.2   57.0    357  4.28   
28          29   0.23  Very Good     D     VS2   60.5   61.0    357  3.96   
29          30   0.23  Very Good     F     VS1   60.9   57.0    357  3.96   
30          31   0.23  Very Good     F     VS1   60.0   57.0    402  4.00   
31          32   0.23  Very Good     F     VS1   59.8   57.0    402  4.04   
32          33   0.23  Very Good     E     VS1   60.7   59.0    402  3.97   
33          34   0.23  Very Good     E     VS1   59.5   58.0    402  4.01   
34          35   0.23  Very Good     D     VS1   61.9   58.0    402  3.92   
35          36   0.23       Good     F     VS1   58.2   59.0    402  4.06   
36          37   0.23       Good     E     VS1   64.1   59.0    402  3.83   
37          38   0.31       Good     H     SI1   64.0   54.0    402  4.29   
38          39   0.26  Very Good     D     VS2   60.8   59.0    403  4.13   
39          40   0.33      Ideal     I     SI2   61.8   55.0    403  4.49   
40          41   0.33      Ideal     I     SI2   61.2   56.0    403  4.49   
41          42   0.33      Ideal     J     SI1   61.1   56.0    403  4.49   
42          43   0.26       Good     D     VS2   65.2   56.0    403  3.99   
43          44   0.26       Good     D     VS1   58.4   63.0    403  4.19   
44          45   0.32       Good     H     SI2   63.1   56.0    403  4.34   
45          46   0.29    Premium     F     SI1   62.4   58.0    403  4.24   
46          47   0.32  Very Good     H     SI2   61.8   55.0    403  4.35   
47          48   0.32       Good     H     SI2   63.8   56.0    403  4.36   
48          49   0.25  Very Good     E     VS2   63.3   60.0    404  4.00   
49          50   0.29  Very Good     H     SI2   60.7   60.0    404  4.33   

       y     z  
0   3.98  2.43  
1   3.84  2.31  
2   4.07  2.31  
3   4.23

In [ ]:
df.tail(10)

Unnamed: 0  carat        cut color clarity  depth  table     price  \
43142       43143   0.46    Premium     D     VS2   60.9   58.0  7.235619   
43143       43144   0.52  Very Good     H     VS2   63.5   58.0  7.236339   
43144       43145   0.52  Very Good     H     VS2   61.4   55.0  7.236339   
43145       43146   0.52  Very Good     E     SI1   62.9   59.0  7.236339   
43146       43147   0.52  Very Good     E     SI1   62.8   56.0  7.236339   
43147       43148   0.52  Very Good     E     SI1   62.7   57.0  7.236339   
43148       43149   0.33      Ideal     F      IF   62.1   56.0  7.236339   
43149       43150   0.56       Fair     E     SI2   64.5   61.0  7.236339   
43150       43151   0.58      Ideal     H     SI1   61.9   59.0  7.236339   
43151       43152   0.52      Ideal     G     SI1   61.3   55.0  7.236339   

          x     y     z  
43142  4.97  4.94  3.02  
43143  5.08  5.09  3.23  
43144  5.19  5.24  3.20  
43145  5.12  5.15  3.23  
43146  5.13  5.15  3.23  
43147  5.10  5.14  3.21  
43148  4.41  4.44  2.75  
43149  5.18  5.14  3.33  
43150  5.33  5.36  3.31  
43151  5.20  5.24  3.20

Заменим цену на ее логарифм

In [ ]:
df['price'] = np.log(df['price'])
df

Unnamed: 0  carat        cut color clarity  depth  table     price  \
0               1   0.23      Ideal     E     SI2   61.5   55.0  5.786897   
1               2   0.21    Premium     E     SI1   59.8   61.0  5.786897   
2               3   0.23       Good     E     VS1    NaN   65.0  5.789960   
3               4   0.29    Premium     I     VS2   62.4   58.0  5.811141   
4               5   0.31       Good     J     SI2   63.3   58.0  5.814131   
...           ...    ...        ...   ...     ...    ...    ...       ...   
43147       43148   0.52  Very Good     E     SI1   62.7   57.0  7.236339   
43148       43149   0.33      Ideal     F      IF   62.1   56.0  7.236339   
43149       43150   0.56       Fair     E     SI2   64.5   61.0  7.236339   
43150       43151   0.58      Ideal     H     SI1   61.9   59.0  7.236339   
43151       43152   0.52      Ideal     G     SI1   61.3   55.0  7.236339   

          x     y     z  
0      3.95  3.98  2.43  
1      3.89  3.84  2.31  
2      4.05  4.07  2.31  
3      4.20  4.23  2.63  
4      4.34  4.35  2.75  
...     ...   ...   ...  
43147  5.10  5.14  3.21  
43148  4.41  4.44  2.75  
43149  5.18  5.14  3.33  
43150  5.33  5.36  3.31  
43151  5.20  5.24  3.20  

[43152 rows x 11 columns]

In [ ]:
df.describe()

Unnamed: 0         carat         depth         table         price  \
count  43152.000000  43152.000000  43081.000000  43152.000000  43152.000000   
mean   21576.500000      0.850243     61.756391     57.483732      7.867556   
std    12457.053745      0.511738      1.421175      2.213001      1.097382   
min        1.000000      0.200000     43.000000     43.000000      5.786897   
25%    10788.750000      0.370000     61.100000     56.000000      6.777647   
50%    21576.500000      0.900000     61.900000     57.000000      8.150468   
75%    32364.250000      1.130000     62.500000     59.000000      8.748503   
max    43152.000000      5.010000     78.200000     95.000000      9.842835   

                  x             y             z  
count  43152.000000  43152.000000  43152.000000  
mean       5.829285      5.831401      3.598703  
std        1.211473      1.229739      0.750341  
min        0.000000      0.000000      0.000000  
25%        4.630000      4.630000      2.840000  
50%        6.060000      6.050000      3.740000  
75%        6.700000      6.690000      4.130000  
max       10.740000     58.900000      8.060000

Просмотрим все возможные значения столбцов

In [ ]:
df["cut"].unique()

array(['Ideal', 'Premium', 'Good', 'Very Good', 'Fair'], dtype=object)

In [ ]:
df["color"].unique()

array(['E', 'I', 'J', 'H', 'F', 'G', 'D'], dtype=object)

In [ ]:
df["clarity"].unique()

array(['SI2', 'SI1', 'VS1', 'VS2', 'VVS2', 'VVS1', 'I1', 'IF'],
      dtype=object)

В колонке depth видим значения NaN. Заменяем их на среднее арифметическое, чтобы не терять данные




In [ ]:
df = df.fillna(df['depth'].mean())

Выкидываем столбец "Unnamed: 0" из-за ненадобности

In [ ]:
df = df.drop(columns=['Unnamed: 0'], axis=1)

In [ ]:
df

carat        cut color clarity      depth  table     price     x     y  \
0       0.23      Ideal     E     SI2  61.500000   55.0  5.786897  3.95  3.98   
1       0.21    Premium     E     SI1  59.800000   61.0  5.786897  3.89  3.84   
2       0.23       Good     E     VS1  61.756391   65.0  5.789960  4.05  4.07   
3       0.29    Premium     I     VS2  62.400000   58.0  5.811141  4.20  4.23   
4       0.31       Good     J     SI2  63.300000   58.0  5.814131  4.34  4.35   
...      ...        ...   ...     ...        ...    ...       ...   ...   ...   
43147   0.52  Very Good     E     SI1  62.700000   57.0  7.236339  5.10  5.14   
43148   0.33      Ideal     F      IF  62.100000   56.0  7.236339  4.41  4.44   
43149   0.56       Fair     E     SI2  64.500000   61.0  7.236339  5.18  5.14   
43150   0.58      Ideal     H     SI1  61.900000   59.0  7.236339  5.33  5.36   
43151   0.52      Ideal     G     SI1  61.300000   55.0  7.236339  5.20  5.24   

          z  
0      2.43  
1      2.31  
2      2.31  
3      2.63  
4      2.75  
...     ...  
43147  3.21  
43148  2.75  
43149  3.33  
43150  3.31  
43151  3.20  

[43152 rows x 10 columns]

In [ ]:
df.select_dtypes(include=['object'])

cut color clarity
0          Ideal     E     SI2
1        Premium     E     SI1
2           Good     E     VS1
3        Premium     I     VS2
4           Good     J     SI2
...          ...   ...     ...
43147  Very Good     E     SI1
43148      Ideal     F      IF
43149       Fair     E     SI2
43150      Ideal     H     SI1
43151      Ideal     G     SI1

[43152 rows x 3 columns]

Надо заинкодить эти 3 колонки с помощью One Hot encoding

In [ ]:
new_df = df.copy()
new_df

carat        cut color clarity      depth  table     price     x     y  \
0       0.23      Ideal     E     SI2  61.500000   55.0  5.786897  3.95  3.98   
1       0.21    Premium     E     SI1  59.800000   61.0  5.786897  3.89  3.84   
2       0.23       Good     E     VS1  61.756391   65.0  5.789960  4.05  4.07   
3       0.29    Premium     I     VS2  62.400000   58.0  5.811141  4.20  4.23   
4       0.31       Good     J     SI2  63.300000   58.0  5.814131  4.34  4.35   
...      ...        ...   ...     ...        ...    ...       ...   ...   ...   
43147   0.52  Very Good     E     SI1  62.700000   57.0  7.236339  5.10  5.14   
43148   0.33      Ideal     F      IF  62.100000   56.0  7.236339  4.41  4.44   
43149   0.56       Fair     E     SI2  64.500000   61.0  7.236339  5.18  5.14   
43150   0.58      Ideal     H     SI1  61.900000   59.0  7.236339  5.33  5.36   
43151   0.52      Ideal     G     SI1  61.300000   55.0  7.236339  5.20  5.24   

          z  
0      2.43  
1      2.31  
2      2.31  
3      2.63  
4      2.75  
...     ...  
43147  3.21  
43148  2.75  
43149  3.33  
43150  3.31  
43151  3.20  

[43152 rows x 10 columns]

In [ ]:
colums_names = list(new_df.select_dtypes(include=['object']))

In [ ]:
df_fin = pd.get_dummies(new_df, columns=colums_names, drop_first=True).copy()
df_fin

carat      depth  table     price     x     y     z  cut_Good  \
0       0.23  61.500000   55.0  5.786897  3.95  3.98  2.43         0   
1       0.21  59.800000   61.0  5.786897  3.89  3.84  2.31         0   
2       0.23  61.756391   65.0  5.789960  4.05  4.07  2.31         1   
3       0.29  62.400000   58.0  5.811141  4.20  4.23  2.63         0   
4       0.31  63.300000   58.0  5.814131  4.34  4.35  2.75         1   
...      ...        ...    ...       ...   ...   ...   ...       ...   
43147   0.52  62.700000   57.0  7.236339  5.10  5.14  3.21         0   
43148   0.33  62.100000   56.0  7.236339  4.41  4.44  2.75         0   
43149   0.56  64.500000   61.0  7.236339  5.18  5.14  3.33         0   
43150   0.58  61.900000   59.0  7.236339  5.33  5.36  3.31         0   
43151   0.52  61.300000   55.0  7.236339  5.20  5.24  3.20         0   

       cut_Ideal  cut_Premium  ...  color_H  color_I  color_J  clarity_IF  \
0              1            0  ...        0        0        0           0   
1              0            1  ...        0        0        0           0   
2              0            0  ...        0        0        0           0   
3              0            1  ...        0        1        0           0   
4              0            0  ...        0        0        1           0   
...          ...          ...  ...      ...      ...      ...         ...   
43147          0            0  ...        0        0        0           0   
43148          1            0  ...        0        0        0           1   
43149          0            0  ...        0        0        0           0   
43150          1            0  ...        1        0        0           0   
43151          1            0  ...        0        0        0           0   

       clarity_SI1  clarity_SI2  clarity_VS1  clarity_VS2  clarity_VVS1  \
0                0            1            0            0             0   
1                1            0            0            0             0   
2                0            0            1            0             0   
3                0            0            0            1             0   
4                0            1            0            0             0   
...            ...          ...          ...          ...           ...   
43147            1            0            0            0             0   
43148            0            0            0            0             0   
43149            0            1            0            0             0   
43150            1            0            0            0             0   
43151            1            0            0            0             0   

       clarity_VVS2  
0                 0  
1                 0  
2                 0  
3                 0  
4                 0  
...             ...  
43147             0  
43148             0  
43149             0  
43150             0  
43151             0  

[43152 rows x 24 columns]

Дальше будет работать с df_fin

In [ ]:
X = df_fin[[name for name in df_fin.columns if name != 'price']]
y = df_fin['price']

In [ ]:
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression

Разбивка данных на тренировочные и тестовые

In [ ]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=13)

In [ ]:
y_test

9922     8.453188
42043    7.145196
23237    9.327946
38955    6.192362
21258    9.149316
           ...   
10378    8.471777
35196    6.792344
39607    6.196444
1748     8.019942
20484    9.084777
Name: price, Length: 8631, dtype: float64

обучение модели

In [ ]:
reg = LinearRegression()

In [ ]:
reg.fit(X_train, y_train)

LinearRegression()

In [ ]:
coeff_df = pd.DataFrame(data=reg.coef_, index=X_train.columns, columns=['Coefficient'])
coeff_df

Coefficient
carat            -0.577652
depth             0.045247
table             0.009705
x                 1.076725
y                 0.024558
z                 0.177465
cut_Good          0.093212
cut_Ideal         0.160944
cut_Premium       0.116684
cut_Very Good     0.130318
color_E          -0.058721
color_F          -0.090357
color_G          -0.158156
color_H          -0.261025
color_I          -0.393700
color_J          -0.540906
clarity_IF        1.124920
clarity_SI1       0.635956
clarity_SI2       0.461201
clarity_VS1       0.848067
clarity_VS2       0.785419
clarity_VVS1      1.035897
clarity_VVS2      0.963484

In [ ]:
y_pred = reg.predict(X_test)

Оценка результата

In [ ]:
print(f'training r-squared {reg.score(X_train, y_train)}')
print(f'testing r-squared {reg.score(X_test, y_test)}')

training r-squared 0.9705735055377107
testing r-squared 0.9817324080481321


Таблица с реальными и запредикченными ценами




In [ ]:
pd.DataFrame({'Actual': y_test, 'Predicted': y_pred})



Actual  Predicted
9922   8.453188   8.371082
42043  7.145196   7.123553
23237  9.327946   9.295004
38955  6.192362   6.364444
21258  9.149316   9.258759
...         ...        ...
10378  8.471777   8.255388
35196  6.792344   6.830298
39607  6.196444   6.247321
1748   8.019942   8.033576
20484  9.084777   9.132025

[8631 rows x 2 columns]

In [ ]:
from sklearn import metrics
print('Mean Absolute Error:', metrics.mean_absolute_error(y_test, y_pred))
print('Mean Squared Error:', metrics.mean_squared_error(y_test, y_pred))
print('Root Mean Squared Error:', np.sqrt(metrics.mean_squared_error(y_test, y_pred)))

Mean Absolute Error: 0.11596364003809025
Mean Squared Error: 0.021967516087395153
Root Mean Squared Error: 0.1482144260434697


Score модели 0.97. Абсолютная и квадратичная ошибки маленькие. Значит модель работает корректно.